logic/LogicManager.java

```diff
@@ -130,34 +130,45 @@ private Rule fromStructure(RuleStructure ruleStructure) {
     }
 
     private void validateCyclesThroughNegations(ConceptManager conceptMgr, LogicManager logicMgr) {
+        // Problem: When a derivation of an inferred concept contains its own negation.
+        // Tracking this at runtime is impractical - statically detect & disallow.
+        // Detect: A cycle which passes through atleast one negation.
+        //      Equivalently - a cycle starting from a negated concept in a rule.
         Set<Rule> negationRulesTriggeringRules = logicMgr.rulesWithNegations()
                 .filter(rule -> !rule.condition().negatedConcludablesTriggeringRules(conceptMgr, logicMgr).isEmpty())
                 .toSet();
 
         for (Rule negationRule : negationRulesTriggeringRules) {
             Map<Rule, RuleDependency> visitedDependentRules = new HashMap<>();
-            visitedDependentRules.put(negationRule, RuleDependency.of(negationRule, null));
-            List<RuleDependency> frontier = new LinkedList<>(ruleDependencies(negationRule, conceptMgr, logicMgr));
-            RuleDependency dependency;
+            LinkedList<RuleDependency> frontier = new LinkedList<>(negatedRuleDependencies(negationRule, conceptMgr, logicMgr));
```

Comment:
interesting - so this indirectly relies on the unifier having pruned the rule from being recursive because the `true` and `false` don't match up right?